ucm2/MediaTek/mt8195_demo/mt8195_demo.conf

```diff
@@ -16,12 +16,12 @@ BootSequence [
 	cset "name='PGA_L_Mux' AIN1"
 	cset "name='HDMI_OUT_MUX' Connect"
 	cset "name='DPTX_OUT_MUX' Connect"
+	# we need to enable all devices before starting PA.
+	# In our driver we use PCM, which means that we have
+	# to route Front End to a BackEnd and then only we can
+	# open a device. Without linking to BE it will fail.
```

Comment:
If you turn a specific UCM device off, can it be re-opened (is the route between frontend / backend preserved)? I mean if the 'off' in DisableSequence for devices are ok or not.

Reply:
If I turn a specific UCM device off in` mt8195_demo.conf ` I can re-opened the device using     ` alsaucm  -c hw:0  set _verb HiFi set _enadev ` .
(This only works if I don't add "disdevall" in **HiFi.conf** )

Reply:
It makes me think that the enable/disable logic is completely broken. It should be allowed:

`set _verb HiFi`
`set _enadev <some_dev>`
`set _enadev <other_dev>`
... the associated PCM devices are availabe for both UCM devices ..
`set _disdev <some_dev>`
... the associated PCM devices are availabe for <other_dev> UCM device ..
`set _ena_dev <some_dev>`
`set _verb Inactive`    # disable verb
`set _verb HiFi` ...

Reply:
I don't follow you. The sequences should be executed. Just run the suggested tests and check the PCM availability (using any ALSA native tool like `aplay/arecord`). The ALSA control elements may be monitored using `amixer events` or `alsactl monitor` commands.

Reply:
Here my tests results: 
```

i1200-demo:~$ set _verb HiFi
i1200-demo:~$ set _enadev Headphones
i1200-demo:~$ set _enadev HDMI      
i1200-demo:~$ set _disdev Headphones
i1200-demo:~$ set _enadev Headphones
i1200-demo:~$ set _verb Inactive
i1200-demo:~$ set _verb HiFi
i1200-demo:~$ arecord -f S16_LE test.wav
Recording WAVE 'test.wav' : Signed 16 bit Little Endian, Rate 8000 Hz, Mono
^CAborted by signal Interrupt...
i1200-demo:~$ aplay test.wav 
Playing WAVE 'test.wav' : Signed 16 bit Little Endian, Rate 8000 Hz, Mono

```

I recorded the sound successfully and was able to play it and listening to my recording successfully using my Headphones.

Reply:
You check pulseaudio PCM/mixer not ALSA PCM/mixer. The `amixer events` should be run simultaneously (all time) on another terminal - it's just for a confirmation that ucm set the right controls. Use '-D hw:0,X' device for arecord/aplay (like it's defined in the UCM config). The aplay/arecord tests should be executed after each 'ucm set' step to check the usability of the PCM devices.

Reply:
Also note that UCM expects to use those PCM devices:

```
alsa-ucm-conf]$ grep PCM ucm2/MediaTek/mt8195_demo/*.conf
ucm2/MediaTek/mt8195_demo/HiFi.conf:		PlaybackPCM "hw:${CardId},5"
ucm2/MediaTek/mt8195_demo/HiFi.conf:		PlaybackPCM "hw:${CardId},0"
ucm2/MediaTek/mt8195_demo/HiFi.conf:		CapturePCM "hw:${CardId},10"
ucm2/MediaTek/mt8195_demo/HiFi.conf:		CapturePCM "hw:${CardId},11"
ucm2/MediaTek/mt8195_demo/HiFi.conf:		PlaybackPCM "hw:${CardId},6"
ucm2/MediaTek/mt8195_demo/HiFi.conf:		CapturePCM "hw:${CardId},14"
```

Reply:
I checked PCM devices (exept hw:0,5 because I don't have a monitor)  after each ucm set (please check the attachment). 
[check PCM devices.txt](https://github.com/alsa-project/alsa-ucm-conf/files/10022976/check.PCM.devices.txt)

Reply:
Hello @perexg, 
I rechecked the usability  of all PCM  devices  and it looks good for me. Can you please confirm? I put the results in the attachment bellow. 
[PCM devices  usability.txt](https://github.com/alsa-project/alsa-ucm-conf/files/10032715/PCM.devices.usability.txt)

Thanks for your support !

Reply:
Which controls makes the PCM devices unavailable ?

Reply:
Hello @perexg,


>  Which controls make the PCM devices unavailable ?
Controls which set to off in DisableSequence.  If we take as an exemple Headphones, **'O176 I070 Switch' and 'O177 I071 Switch'** makes PCM device **hw:0,0** unavailable if they are setting to OFF. 
  
FYI, I set the Switch to OFF using amixer ( amixer sset -c0 <switch-id> off), because when calling DisableSequence the stat of the Switch does not change (it's like ` alsaucm  -c hw:0  set _verb HiFi set _disdev <device_name>`  does nothing).

Reply:
DisableSequence is set only for the active (previously enabled devices). So you should use _enadev and _disdev...

Reply:
> Controls which set to off in DisableSequence. If we take as an exemple Headphones, 'O176 I070 Switch' and 'O177 I071 Switch' makes PCM device hw:0,0 unavailable if they are setting to OFF.

So with this change, the UCM application won't be able to use the PCM device next time. Basically, because there are no ConflictingDevices in your UCM configuration, PA opens all devices together, so the problem is not visible.  I think that the controls which affects the PCM device availability should be set only in verb's Enable/Disable sequences to make sure that they are live for the whole time when the UCM verb is used.

Reply:
So my recommendation is:

1) remove off from the BootSequence
2) add all PCM related controls also to the verb sequences (Enable/Disable)
3) do not use PCM related controls in the device configuration blocks

Reply:
Thanks @perexg !